src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,257 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Orleans.Runtime.Configuration.Options;
+
+namespace Orleans.Runtime.Placement;
+
+internal sealed class ResourceOptimizedPlacementDirector : IPlacementDirector, ISiloStatisticsChangeListener
+{
+    readonly record struct ResourceStatistics(float? CpuUsage, float? AvailableMemory, long? MemoryUsage, long? TotalPhysicalMemory, bool IsOverloaded);
+
+    /// <summary>
+    /// 1 / (1024 * 1024)
+    /// </summary>
+    private const float PhysicalMemoryScalingFactor = 0.00000095367431640625f;
+    
+    private readonly ResourceOptimizedPlacementOptions _options;
+    private readonly ConcurrentDictionary<SiloAddress, ResourceStatistics> siloStatistics = [];
+
+    private readonly DualModeKalmanFilter<float> _cpuUsageFilter = new();
+    private readonly DualModeKalmanFilter<float> _availableMemoryFilter = new();
+    private readonly DualModeKalmanFilter<long> _memoryUsageFilter = new();
+
+    private Task<SiloAddress> _cachedLocalSilo;
+
+    public ResourceOptimizedPlacementDirector(
+        DeploymentLoadPublisher deploymentLoadPublisher,
+        IOptions<ResourceOptimizedPlacementOptions> options)
+    {
+        _options = options.Value;
+        deploymentLoadPublisher?.SubscribeToStatisticsChangeEvents(this);
+    }
+
+    public Task<SiloAddress> OnAddActivation(PlacementStrategy strategy, PlacementTarget target, IPlacementContext context)
+    {
+        var compatibleSilos = context.GetCompatibleSilos(target);
+
+        if (IPlacementDirector.GetPlacementHint(target.RequestContextData, compatibleSilos) is { } placementHint)
+        {
+            return Task.FromResult(placementHint);
+        }
+
+        if (compatibleSilos.Length == 0)
+        {
+            throw new SiloUnavailableException($"Cannot place grain with Id = [{target.GrainIdentity}], because there are no compatible silos.");
+        }
+
+        if (compatibleSilos.Length == 1)
+        {
+            return Task.FromResult(compatibleSilos[0]);
+        }
+
+        if (siloStatistics.IsEmpty)
+        {
+            return Task.FromResult(compatibleSilos[Random.Shared.Next(compatibleSilos.Length)]);
+        }
+
+        var bestCandidate = GetBestSiloCandidate(compatibleSilos);
+
+        if (IsLocalSiloPreferable(context, compatibleSilos, bestCandidate.Value))
+        {
+            return _cachedLocalSilo ??= Task.FromResult(context.LocalSilo);
+        }
+
+        return Task.FromResult(bestCandidate.Key);
+    }
+
+    private KeyValuePair<SiloAddress, float> GetBestSiloCandidate(SiloAddress[] compatibleSilos)
+    {
+        List<KeyValuePair<SiloAddress, ResourceStatistics>> relevantSilos = [];
```

Comment:
Performance-wise I think this implementation is ok for now. We could consider using stackalloc'd spans for some of the collections, falling back to heap allocated as needed, and avoiding the allocation-happy LINQ usage.

Reply:
@ReubenBond have at look at `GetBestSiloCandidate_V2` and let me know